### PR TITLE
Adding support for the new footer

### PR DESCRIPTION
**Pull request by ahmad-shahid**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,15 @@
...
 
+## v2.8.0
+
+### New features and improvements
+
+- **IMPORTANT** The GCWeb site footer has been updated to reflect the changes introduced in WET footer version 4. These changes will be applied automatically. For more information, please visit the WET documentation: https://wet-boew.github.io/GCWeb/sites/footers/footers-en.html
+- For information on how to implement the new features introduced with the footer, please visit our sample pages: https://github.com/wet-boew/cdts-DotNetTemplates/tree/master/samples
+- Updating static fallback files
+- [CDTS](https://gccode.ssc-spc.gc.ca/iitb-dgiit/nw-ws/sgdc-cdts) v4.0.47 & [wet-boew](https://github.com/wet-boew/wet-boew) v4.0.56.5
+- Internal changes and bug fixes
```
Comment (plup2) [inline CHANGELOG.md:11]: "Internal changes" was more for SOY->EJS and nodocwrite changes, we can stick with "Bug fixes" for the "normal releases"

File: Components/Entities/Footer.cs [hunk 2]
```diff
@@ -7,12 +7,16 @@ public class Footer
...
         public bool ShowFooter { get; set; }
         public List<Link> ContactLinks { get; set; }
-        public List<FooterLink> PrivacyLink { get; set; }
-        public List<FooterLink> TermsLink { get; set; } 
         public string LocalPath { get; set; }
+        public bool HideFooterMain { get; set; }
+        public bool HideFooterCorporate { get; set; }
+        public ContextualFooter ContextualFooter { get; set; }
+        [JsonProperty("privacyLink")]
+        public SubFooterLink PrivacyFooterLink { get; set; }
```
Comment (plup2) [inline Components/Entities/Footer.cs:18]: We'll need to talk about that one :)

File: Components/Entities/TransactionalFooter.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+﻿using Newtonsoft.Json;
```
Comment (plup2) [inline Components/Entities/TransactionalFooter.cs:1]: This class seems superfluous, we'll talk :)

File: Components/Entities/SubFooterLink.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+
+namespace GoC.WebTemplate.Components.Entities
+{
+    public class SubFooterLink
```
Comment (plup2) [inline Components/Entities/SubFooterLink.cs:5]: Let's talk about that one too...

File: Components/IModel.cs [hunk 5]
```diff
@@ -216,6 +241,12 @@ public interface IModel
         /// </summary>
         FooterLink TermsConditionsLink { get; set; }
 
+        /// <summary>
+        /// Configures the Terms and Conditions Link in the footer
+        /// Set by application programmatically
+        /// </summary>
+        SubFooterLink TermsFooterLink { get; set; }
```
Comment (plup2) [inline Components/IModel.cs:248]: Isn't that the same as TermsConditionsLink?

File: CHANGELOG.md (hunk 1 above)
Comment (ahmad-shahid) [reply CHANGELOG.md]: done

File: Components/Entities/TransactionalFooter.cs (hunk 3 above)
Comment (ahmad-shahid) [reply Components/Entities/TransactionalFooter.cs]: Removed

File: Components/Entities/SubFooterLink.cs (hunk 4 above)
Comment (ahmad-shahid) [reply Components/Entities/SubFooterLink.cs]: Removed

File: Components/IModel.cs (hunk 5 above)
Comment (ahmad-shahid) [reply Components/IModel.cs]: Removed

File: Components/Entities/Footer.cs (hunk 2 above)
Comment (ahmad-shahid) [reply Components/Entities/Footer.cs]: Footer class updated

File: Components.Tests/RenderTests/RenderFooterTests.cs [hunk 6]
```diff
@@ -234,5 +234,34 @@ public void HandleContactLinkValue(Model sut)
...
+        }
+
+        [Theory, AutoNSubstituteData]
+        public void ModifyTermsPrivacyLink(Model sut)
+        {
+            sut.PrivacyLink = new FooterLink() { Href = "google" };
+            sut.TermsConditionsLink = new FooterLink() { Href = "google" };
+            sut.Render.Footer().ToString().Should().Contain("\"privacyLink\":{\"href\":\"google\"},\"termsLink\":{\"href\":\"google\"}");
+        }
     }
```
Comment (plup2) [inline Components.Tests/RenderTests/RenderFooterTests.cs:266]: Should have a test to verify the switch between list and single depending on showFooter

File: Components/Entities/Footer.cs [hunk 7]
```diff
@@ -1,18 +1,59 @@
...
+        internal class FooterLinkConverter : JsonConverter
+        {
+            public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+            {
+                var footerLink = value as FooterLinkContext;
+
+                //this converter should never be on a type that's not a FooterLinkContext so just throw exceptions.
+                Debug.Assert(footerLink != null, "The footerLink cannot be null.");
+
+                var listFooterLink = new List<FooterLink> { footerLink.FooterLink };
```
Comment (plup2) [inline Components/Entities/Footer.cs:35]: Should avoid creating object unless they are needed (for performance, although admittedly here it is not a very big deal)
List creation should go in the "if"

File: Components/Entities/FooterLinkContext.cs [hunk 8]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+
+namespace GoC.WebTemplate.Components.Entities
+{
+    public class FooterLinkContext
+    {
+        /**
```
Comment (plup2) [inline Components/Entities/FooterLinkContext.cs:7]: Class comment goes above the class ;)

File: Components/IModel.cs [hunk 9]
```diff
@@ -105,6 +111,19 @@ public interface IModel
         /// </summary>
         bool HidePlaceholderMenu { get; set; }
 
+        /// <summary>
+
```
Comment (plup2) [inline Components/IModel.cs:115]: Would not have mentioned it if it was the only thing but... weird blank line :)

File: Components/Utils/ModelRenderer.cs [hunk 10]
```diff
@@ -177,8 +179,8 @@ public HtmlString TransactionalFooter()
                 SubTheme = _model.CdtsEnvironment.SubTheme,
                 ShowFooter = false,
                 ContactLinks = _model.Builder.BuildContactLinks(),
-                PrivacyLink = _model.Builder.BuildSingleFooterLink(_model.PrivacyLink),
-                TermsLink = _model.Builder.BuildSingleFooterLink(_model.TermsConditionsLink)
+                PrivacyLink = string.IsNullOrEmpty(_model.PrivacyLink.Href) ? null : new FooterLinkContext { ShowFooter = false, FooterLink = _model.PrivacyLink },
```
Comment (plup2) [inline Components/Utils/ModelRenderer.cs:182]: This is ok, but it side-steps function `Builder.BuildSingleFooterLink` which becomes useless and the logic is duplicated 4 times.  I suggest changing the `Builder.BuildSingleFooterLink` to return `FooterLinkContext` instead (and maybe rename the function to match it new goal).   This function should be used here for both links and in `Footer()`.

File: Components/Utils/ModelRenderer.cs [hunk 11]
```diff
@@ -177,8 +179,8 @@ public HtmlString TransactionalFooter()
...
+                TermsLink = string.IsNullOrEmpty(_model.TermsConditionsLink.Href) ? null : new FooterLinkContext { ShowFooter = false, FooterLink = _model.TermsConditionsLink },
```
Comment (plup2) [inline Components/Utils/ModelRenderer.cs:184]: This is not directly related to a specific code change but:   Footer() and TransactionalFooter() are also used with GCIntranet theme, no matter its subtheme, with basic or application template.  Did you confirm that this change of "terms/privacy link is not an array when showFooter is true" also works for gcintranet?

**Review by plup2 — CHANGES_REQUESTED** [review]

File: Components/Utils/ModelRenderer.cs (hunk 11 above)
Comment (plup2) [reply Components/Utils/ModelRenderer.cs:184]: Ah and also, that it is ok with GCweb appFooter

File: Components/IModel.cs (hunk 9 above)
Comment (ahmad-shahid) [reply Components/IModel.cs]: done

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,5 +1,15 @@
 # DotNetTemplates Change Log
 
+## v2.8.0
+
+### New features and improvements
+
+- **IMPORTANT** The GCWeb site footer has been updated to reflect the changes introduced in WET footer version 4. These changes will be applied automatically. For more information, please visit the WET documentation: https://wet-boew.github.io/GCWeb/sites/footers/footers-en.html
+- For information on how to implement the new features introduced with the footer, please visit our sample pages: https://github.com/wet-boew/cdts-DotNetTemplates/tree/master/samples
```
Comment (plup2) [inline CHANGELOG.md:8]: This should be bla bla test

File: CHANGELOG.md [hunk 13]
```diff
@@ -1,5 +1,15 @@
...
+- Updating static fallback files
+- [CDTS](https://gccode.ssc-spc.gc.ca/iitb-dgiit/nw-ws/sgdc-cdts) v4.0.47 & [wet-boew](https://github.com/wet-boew/wet-boew) v4.0.56.5
+- Bug fixes
+
```
Comment (plup2) [inline CHANGELOG.md:12]: This is a second comment, isn't nice? (test once more)

**Review by plup2 — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 12 above)
Comment (ahmad-shahid) [reply CHANGELOG.md:8]: done

File: CHANGELOG.md (hunk 13 above)
Comment (ahmad-shahid) [reply CHANGELOG.md:12]: done

File: Components.Tests/RenderTests/RenderFooterTests.cs (hunk 6 above)
Comment (plup2) [reply Components.Tests/RenderTests/RenderFooterTests.cs:274]: This is good, but add comments in each of the last two tests to explain what is it _exactly_ that is being tested.

File: Components/Utils/ModelRenderer.cs (hunk 11 above)
Comment (plup2) [reply Components/Utils/ModelRenderer.cs:184]: will and not should, perfect

File: Components.Tests/RenderTests/RenderFooterTests.cs (hunk 6 above)
Comment (ahmad-shahid) [reply Components.Tests/RenderTests/RenderFooterTests.cs:280]: done

**Review by plup2 — APPROVED** [review]
